### PR TITLE
WIP - Semeru Compiler Service

### PR DESCRIPTION
**Pull request by yongja79**
Signed-off-by: Jason Yong <jason_yong@uk.ibm.com>

### CODE REVIEW

File: controllers/semeru_compiler.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// Create the Deployment and Service objects for a Semeru Compiler used by a Websphere Liberty Application
+func (r *ReconcileWebSphereLiberty) reconcileSemeruCompiler(la *wlv1.WebSphereLibertyApplication) (error, string) {
+	compilerMeta := metav1.ObjectMeta{
+		Name:      la.GetName() + `-semeru-compiler`,
+		Namespace: la.GetNamespace(),
+	}
+	// Create the Semeru Service object
+	semsvc := &corev1.Service{ObjectMeta: compilerMeta}
+	err := r.CreateOrUpdate(semsvc, la, func() error {
+		createSemeruService(semsvc, la)
```
Comment (leochr) [inline controllers/semeru_compiler.go:18]: Suggest method name since it could also be updated: reconcileSemeruService

**Review by leochr — COMMENTED** [review]
@yongja79 Thanks for the PR! Looks good for the most part. Added some comments.

File: controllers/semeru_compiler.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	deploy.Spec.Template = corev1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   wlva.Name + SemeruLabelName,
+			Labels: getLabels(wlva),
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:            JitServer,
+					Image:           ImageURL,
```
Comment (leochr) [inline controllers/semeru_compiler.go:84]: The application image should be used here. Refer to the Deployment created for the application on how image is referenced. I believe the image detail is stored in status.

File: controllers/semeru_compiler.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+					ImagePullPolicy: "IfNotPresent",
```
Comment (leochr) [inline controllers/semeru_compiler.go:85]: Use the same pull policy as application since the same application image is used

**Review by leochr — COMMENTED** [review]
@yongja79 Thanks for addressing the review comments.

@smcclem Thanks for the Deployment related changes. Using this PR as it has the latest code. Looks good for the most part. Some changes related to the image are needed.

File: controllers/semeru_compiler.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+					Command:         []string{"jitserver"},
+					Ports: []corev1.ContainerPort{
+						{
+							ContainerPort: 8443,
```
Comment (leochr) [inline controllers/semeru_compiler.go:89]: @smcclem This port should be 38400, to match the port specified in Service

**Comment by smcclem** [comment]
@leochr @yongja79  I've taken these files and made my code updates there :  See https://github.com/WASdev/websphere-liberty-operator/pull/299  

We can talk about it on the scrum tomorrow, but I am having an issue in testing.   The env: name value pair I am adding for the OpenJ9 arguments are being nulled out in the (application) deployment:   

`apiVersion: liberty.websphere.ibm.com/v1
kind: WebSphereLibertyApplication
metadata:
  name: websphereliberty-app-sample
  namespace: semeru-test
spec:
  license:
    accept: true
    edition: IBM WebSphere Application Server
    productEntitlementSource: Standalone
    metric: Virtual Processor Core (VPC)
  applicationImage: XXX.com/semeru-test/petclinic
  env:
  # Provide additional options to the JVM
  - name: OPENJ9_JAVA_OPTIONS
  value: "-XX:+UseJITServer -XX:JITServerAddress=websphereliberty-app-sample-semeru-compiler.semeru-test.svc.cluster.local -XX:+JITServerLogConnections"
`

**Review by leochr — APPROVED** [review]
Looks good. Thank you both!

**Comment by leochr** [comment]
@smcclem I believe the indentation of `value` is not correct. The `value` must be missing from the processed WebSphereLibertyApplication CR as well. It should be added like this:

```
  env:
  # Provide additional options to the JVM
  - name: OPENJ9_JAVA_OPTIONS
    value: "-XX:+UseJITServer -XX:JITServerAddress=websphereliberty-app-sample-semeru-compiler.semeru-test.svc.cluster.local -XX:+JITServerLogConnections"
```

**Pull request merged**